File: leetcode_java/src/main/java/com/programapprentice/app/MaximumProductSubarray.java
package com.programapprentice.app;

/**
 * Created with IntelliJ IDEA.
 * User: ruijiang
 * Date: 10/26/14
 * Time: 10:20 PM
 */

/**
 * Find the contiguous subarray within an array (containing at least one number) which has the largest product.
 * For example, given the array [2,3,-2,4],
 * the contiguous subarray [2,3] has the largest product = 6.
 * */
public class MaximumProductSubarray {

    // brute force method O(n^2)
    public int maxProductV1(int[] A) {
        if(A == null || A.length == 0) {
            return 0;
        }
        int n = A.length;
        int[][] scan = new int[n][n];
        int max = Integer.MIN_VALUE;
        for(int i = 0; i < n; i++) {
            scan[i][i] = A[i];
            max = max > scan[i][i] ? max : scan[i][i];
            for(int j = i+1; j < n; j++) {
                scan[i][j] = scan[i][j-1] * A[j];
                max = max > scan[i][j] ? max : scan[i][j];
            }
        }

        return max;
    }

    public int maxProduct(int[] A) {
        if(A == null || A.length == 0) {
            return 0;
        }
        int max = A[0];
        int min = A[0];
        int globalMax = A[0];
        int tmpMax = max;
        int tmpMin = min;
        for(int i = 1; i < A.length; i++) {
            tmpMax = max;
            tmpMin = min;
            max = max3(tmpMax * A[i], tmpMin*A[i], A[i]);
            min = min3(tmpMax * A[i], tmpMin*A[i], A[i]);
            globalMax = Math.max(globalMax, max);
        }
        return globalMax;
    }

    public int max3(int a, int b, int c) {
        return Math.max(a, Math.max(b, c));
    }

    public int min3(int a, int b, int c) {
        return Math.min(a, Math.min(b, c));
    }

    // http://blog.csdn.net/worldwindjp/article/details/39826823
    public int maxProductSolution(int[] A) {
        if(A == null || A.length == 0) {
            return 0;
        }
        int max_local = A[0];
        int min_local = A[0];
        int global = A[0];

        for(int i=1;i<A.length;i++)
        {
            int max_copy = max_local;
            max_local = Math.max(Math.max(A[i]*max_local, A[i]), A[i]*min_local);
            min_local = Math.min(Math.min(A[i]*max_copy, A[i]), A[i]*min_local);
            global = Math.max(global, max_local);
        }
        return global;
    }
}

File: leetcode_cpp/src/maximum_product_subarray.cc
/*
 * maximum_product_subarray.cc
 *
 *  Created on: Dec 10, 2014
 *      Author: ruijiang
 */

#include <climits>
#include "maximum_product_subarray.h"
#include <iostream>

using namespace::std;

int Solution::max3(int a, int b, int c) {
    int m = a > b ? a : b;
    m = m > c ? m : c;
    return m;
}

int Solution::min3(int a, int b, int c) {
    int m = a > b ? b : a;
    m = m > c ? c : m;
    return m;
}

int Solution::maxProduct(int A[], int n) {
    if(0 == n) {
        return 0;
    }
    int min = A[0];
    int max = A[0];
    int tmin = min, tmax = max;
    int global_max = A[0];
    int tmp1, tmp2;
    for(int i = 1; i < n; i++) {
        tmp1 = min * A[i];
        tmp2 = max * A[i];
//        cout << "tmp1 " << tmp1 << " tmp2 " << tmp2 << " A[i] " << A[i] << endl;
        min = min3(tmp1, tmp2, A[i]);
        max = max3(tmp1, tmp2, A[i]);
        global_max = global_max > max ? global_max : max;
//        cout << "min: " << min << " max: " << max << " global max " << global_max << endl;
    }
    return global_max;
}
